### PR TITLE
fix: make OpenAIMultiModal work with new ChatMessage

### PR DESCRIPTION
**Pull request by masci**
# Description

Fix OpenAIMultiModal so it works properly with the new `ChatMessage` structure.

## Version Bump?

No version bump, this needs to be released after we release the new `ChatMessage` in core.

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] I added new unit tests to cover this change
- [ ] I believe this change is already covered by existing unit tests

### CODE REVIEW

File: llama-index-core/llama_index/core/llms/function_calling.py [hunk 1]
```diff
@@ -27,6 +22,10 @@ class FunctionCallingLLM(LLM):
 
     """
 
+    def __init__(*args, **kwargs) -> None:
+        # Help static checkers understand this class hierarchy
+        super().__init__(*args, **kwargs)
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/llms/function_calling.py:27]: Huh, TIL

File: llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/base.py [hunk 2]
```diff
@@ -964,18 +963,3 @@ def stream_structured_predict(
         # by default structured prediction uses function calling to extract structured outputs
         # here we force tool_choice to be required
         return super().stream_structured_predict(*args, llm_kwargs=llm_kwargs, **kwargs)
-
-    @dispatcher.span
-    def stream_structured_predict(
-        self, *args: Any, llm_kwargs: Optional[Dict[str, Any]] = None, **kwargs: Any
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/base.py:970]: I wonder if this was meant to be astream_structured_predict

File: llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/utils.py [hunk 3]
```diff
@@ -250,10 +255,32 @@ def is_function_calling_model(model: str) -> bool:
 def to_openai_message_dict(
     message: ChatMessage, drop_none: bool = False, model: Optional[str] = None
 ) -> ChatCompletionMessageParam:
-    """Convert generic message to OpenAI message dict."""
+    """Convert a ChatMessage to an OpenAI message dict."""
+    content = []
+    for block in message.blocks:
+        if isinstance(block, TextBlock):
+            content.append({"type": "text", "text": block.text})
+        elif isinstance(block, ImageBlock):
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/utils.py:263]: Should we have an else to raise an error for any other types? Just to catch ourselves in the future

File: llama-index-core/llama_index/core/llms/function_calling.py (hunk 1 above)
Comment (masci) [reply llama-index-core/llama_index/core/llms/function_calling.py:27]: Yeah Pyright gets lost without an explicit __init__ when a pydantic model is part of the hierarchy

File: llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/base.py (hunk 2 above)
Comment (masci) [reply llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/base.py:970]: I couldn't find any usage of this, should I bring it back in async fashion?

File: llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/utils.py (hunk 3 above)
Comment (masci) [reply llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/utils.py:263]: Good idea 👍

File: llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/base.py (hunk 2 above)
Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-openai/llama_index/llms/openai/base.py:970]: There should be a definition in the base `LLM` class that we can copy yea, its basically just here to force the tool choice with openai (streaming structured outputs are definitely under-utilized, but very cool feature)

**Pull request merged**